src/cache-utils.ts

```diff
@@ -66,35 +103,131 @@ export const getPackageManagerInfo = async (packageManager: string) => {
   } else if (packageManager === 'pnpm') {
     return supportedPackageManagers.pnpm;
   } else if (packageManager === 'yarn') {
-    const yarnVersion = await getPackageManagerVersion('yarn', '--version');
-
-    core.debug(`Consumed yarn version is ${yarnVersion}`);
-
-    if (yarnVersion.startsWith('1.')) {
-      return supportedPackageManagers.yarn1;
-    } else {
-      return supportedPackageManagers.yarn2;
-    }
+    return supportedPackageManagers.yarn;
   } else {
     return null;
   }
 };
 
-export const getCacheDirectoryPath = async (
+/**
+ * glob expanding memoized because it involves potentially very deep
+ * traversing through the directories tree
+ */
+export const memoizedCacheDependencies: Record<string, string[]> = {};
+/**
+ * Expands (converts) the string input `cache-dependency-path` to list of directories that
+ * may be project roots
+ * @param cacheDependencyPath - either a single string or multiline string with possible glob patterns
+ *                              expected to be the result of `core.getInput('cache-dependency-path')`
+ * @return list of directories and possible
+ */
+const getProjectDirectoriesFromCacheDependencyPath = async (
+  cacheDependencyPath: string
+): Promise<string[]> => {
+  let cacheDependenciesPaths: string[];
+
+  // memoize unglobbed paths to avoid traversing FS
+  const memoized = memoizedCacheDependencies[cacheDependencyPath];
+  if (memoized) {
+    cacheDependenciesPaths = memoized;
+  } else {
+    cacheDependenciesPaths = (await glob
+      .create(cacheDependencyPath)
+      .then(globber => globber.glob())) || [''];
+    memoizedCacheDependencies[cacheDependencyPath] = cacheDependenciesPaths;
+  }
+
+  const existingDirectories: string[] = cacheDependenciesPaths
+    .map(cacheDependencyPath => path.dirname(cacheDependencyPath))
+    // uniq in order to do not traverse the same directories during the further processing
+    .filter(
+      (cachePath, i, result) =>
+        cachePath != null && result.indexOf(cachePath) === i
```

Comment:
Scanning through `result` like that on each iteration would have O(n^2) time complexity which is very unnecessary here and uniqueness filter is general-purpose enough to deserve it's own declaration. Consider:
```typescript
// somwhere in utils folder or something
const unique = () => {
  const encountered = new Set();
  return (value) => {
    if (encountered.has(value)) return false;
    encountered.add(value);
    return true;
  }
}

/* ... */

// here
.filter(cachePath => cachePath !== null)
.filter(unique())

/* ... */
```

Reply:
We know the array to be iterated in this particular case always will be very tiny - 2 or 3, well max 5 items and iteration over them is chipper or equal than allocation/freeing of the not so simple object and building the binary tree for just few values.